### PR TITLE
objection.transaction.start, commit and rollback for transaction

### PR DESCRIPTION
**Pull request by nbeydoon**
Easier way to run transactions

### CODE REVIEW

**Comment by nbeydoon** [comment]
Erf you check the coverage 0 :(

**Comment by nbeydoon** [comment]
Need more work

File: lib/transaction.js [hunk 1]
```diff
@@ -122,6 +125,33 @@ module.exports = function transaction() {
...
+
+  if (!utils.isSubclassOf(modelBase, Model)) {
+    return Promise.reject(new Error('objection.transaction.start: all but the last argument should be Model subclasses'));
+  }
+
+  var knex = modelBase.knex();
+
+  return new Promise(function(resolve, reject) {
+    knex.transaction(function (trx) {
+        modelBase = modelBase.knex(trx);
```
Comment (koskimas) [inline lib/transaction.js]: This sets the the transaction globally for the model. Consider the case that 100 requests are currently running on your node server that use the `modelBase` class. You set the knex connection of _all_ of them here. Very bad things will happen. This line is actually useless.

File: lib/transaction.js [hunk 2]
```diff
@@ -122,6 +125,33 @@ module.exports = function transaction() {
...
+
+      resolve({
```
Comment (koskimas) [inline lib/transaction.js]: You resolve the promise with a custom object. You cannot bind models to this object using `bindKnex`. You can just resolve the promise with the `trx` object which is itself a knex instance.

File: lib/transaction.js [hunk 3]
```diff
@@ -122,6 +125,33 @@ module.exports = function transaction() {
...
+        commit: function() {
+          return trx.commit();
+        },
+        rollback: function() {
+          return trx.rollback();
+        }
+      });
+    }).catch(reject);
```
Comment (koskimas) [inline lib/transaction.js]: Passing the reject method for `cathch` here never does anything. To use the transaction the promise has to already be resolved (with the transaction object). The resolved promise cannot change its state. Therefore the `catch(reject)` does nothing.

Comment (nbeydoon) [inline lib/transaction.js]: So when I run without the catch One 1 / 2 time i get timer error from bluebird, but I agree I don 't understand why

File: lib/transaction.js (hunk 1 above)
Comment (nbeydoon) [inline lib/transaction.js]: Yes I know, I did it like this for my test, and I have wait your review, I think is possible to do similar think like the ````.transaction```` method.

**Comment by koskimas** [comment]
I just pushed a commit to master that should solve this.

**Comment by nbeydoon** [comment]
Yes, it works great but so like this all knex are binded or not?

**Comment by nbeydoon** [comment]
I works great! :)

**Comment by koskimas** [comment]
No, only the knex that is returned by the start function. If you really want to bind all the models to the same transaction (very dangerous and extremely error-prone) you can just say `objection.Model.knex(transaction)`. Note that the transaction object is just like any knex instance. It is just bound to a single connection that has the `BEGIN` command ran.

**Comment by nbeydoon** [comment]
Ok I see, Its ok now I can do all my test and you have a new cool features for objection, very thanks :)

**Comment by koskimas** [comment]
Thank you! This is a cool feature. I just never needed it myself and didn't think of it. Keep the suggestions coming and open issues whenever you have a problem.

**Pull request closed without merging**